include/nbl/builtin/hlsl/emulated_float64_t_utils.hlsl

```diff
@@ -0,0 +1,365 @@
+        ef64_t3x3 lhsTransposed = getTransposed();
+
+        output.columns[0].x = (lhsTransposed.columns[0] * rhs.columns[0]).calcComponentSum();
+        output.columns[0].y = (lhsTransposed.columns[0] * rhs.columns[1]).calcComponentSum();
+        output.columns[0].z = (lhsTransposed.columns[0] * rhs.columns[2]).calcComponentSum();
+
+        output.columns[1].x = (lhsTransposed.columns[1] * rhs.columns[0]).calcComponentSum();
+        output.columns[1].y = (lhsTransposed.columns[1] * rhs.columns[1]).calcComponentSum();
+        output.columns[1].z = (lhsTransposed.columns[1] * rhs.columns[2]).calcComponentSum();
+
+        output.columns[2].x = (lhsTransposed.columns[2] * rhs.columns[0]).calcComponentSum();
+        output.columns[2].y = (lhsTransposed.columns[2] * rhs.columns[1]).calcComponentSum();
+        output.columns[2].z = (lhsTransposed.columns[2] * rhs.columns[2]).calcComponentSum();
+
+        // TODO: avoid transpose
+        return output.getTransposed();
+    }
+
+    ef64_t3 operator*(NBL_CONST_REF_ARG(ef64_t3) rhs)
+    {
+        ef64_t3 output;
+        ef64_t3x3 lhsTransposed = getTransposed();
+
+        output.x = (columns[0] * rhs).calcComponentSum();
+        output.y = (columns[1] * rhs).calcComponentSum();
+        output.z = (columns[2] * rhs).calcComponentSum();
+
+        return output;
+    }
+};
+
+struct ef64_t2x2
+{
+    ef64_t2 columns[2];
+
+    ef64_t2x2 getTransposed() NBL_CONST_MEMBER_FUNC
+    {
+        ef64_t2x2 output;
+
+        output.columns[0].x = columns[0].x;
+        output.columns[1].x = columns[0].y;
+
+        output.columns[0].y = columns[1].x;
+        output.columns[1].y = columns[1].y;
+
+        return output;
+    }
+
+    ef64_t2x2 operator*(NBL_CONST_REF_ARG(ef64_t2x2) rhs) NBL_CONST_MEMBER_FUNC
+    {
+        ef64_t2x2 output;
+        ef64_t2x2 lhsTransposed = getTransposed();
+
+        output.columns[0].x = (lhsTransposed.columns[0] * rhs.columns[0]).calcComponentSum();
+        output.columns[0].y = (lhsTransposed.columns[0] * rhs.columns[1]).calcComponentSum();
+
+        output.columns[1].x = (lhsTransposed.columns[1] * rhs.columns[0]).calcComponentSum();
+        output.columns[1].y = (lhsTransposed.columns[1] * rhs.columns[1]).calcComponentSum();
+
+        return output.getTransposed();
+    }
+
+    ef64_t2 operator*(NBL_CONST_REF_ARG(ef64_t2) rhs)
+    {
+        ef64_t2 output;
+        ef64_t2x2 lhsTransposed = getTransposed();
+
+        output.x = (columns[0] * rhs).calcComponentSum();
+        output.y = (columns[1] * rhs).calcComponentSum();
+
+        return output;
+    }
+};
+
+#endif
+
+// struct VecT is solution to
+// error: 'nbl::hlsl::emulated_float64_t<false, true>' cannot be used as a type parameter where a scalar is required
+// using float_t2 = typename conditional<is_same<float_t, emulated_float64_t<false, true> >::value, ef64_t2, vector<float_t, 2> >::type;
+
+// TODO: better solution
+
+#ifndef __HLSL_VERSION
+using F64_t = double;
```

Comment:
do this
```cpp
template<typename device_capabilities=void>
using portable_float64_t = conditional_t<device_capability_traits<device_capabilities>::shaderFloat64,float64_t,emulated_float64_t>;

// because only structs can partially specialize
namespace impl
{
template<typename T, uint32_t N, bool native=is_scalar_v<T> >
struct portable_vector
{
   using type = emulated_vector<T,N>;
};
// specialization for builtins
template<typename T, uint32_t N>
struct portable_vector<T,N,true>
{
   using type = vector<T,N>;
};
}
template<typename T, uint32_t N>
using portable_vector_t = impl::portable_vector<T,N>::type;
```

A `portable_matrix`  I leave as an exercise to the reader.

Reply:
cool solution